File: reqworker.py
import requests
import time

from utils import timestamp, time_to_str, elapsed_str
from process_data import process_data_and_draw_ui
from colors import bool_to_ctext_fg, int_to_ctext_fg, ctext_fg, COLOR_ON, COLOR_OFF
from logger_config import requests_logger, changes_logger

GLOBAL_UNIX_COUNTER = int(time.time() * 1000)

AUTO_TIMER = True
AUTO_TIMER_STARTED = False
AUTO_TIMER_SECONDS = 900  # 15 mins
AUTO_TIMER_SECONDS_ELAPSED = 0
AUTO_TIMER_STATUS = ""
AUTO_GAS = True
AUTO_GAS_STATUS = ""

HISTORY_MODE = None
HISTORY_MODE_TIME_CHANGED = None
HISTORY_MODE_TIME_ON = None
HISTORY_MODE_TIME_OFF = None
HISTORY_GAS = None
HISTORY_GAS_TIME_CHANGED = None
HISTORY_GAS_TIME_ON = None
HISTORY_GAS_TIME_OFF = None

# act like firefox
req_headers = {
    "Accept": "application/json, text/javascript, */*; q=0.01",
    "Accept-Encoding": "gzip, deflate",
    "Accept-Language": "en-US,en;q=0.5",
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
    "Referer": "http://192.168.1.250/",
    "Sec-GPC": "1",
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:135.0) Gecko/20100101 Firefox/135.0",
    "X-Requested-With": "XMLHttpRequest",
}

URLS = {
    "VARS": "http://192.168.1.250/isc/get_var_js.aspx?StatusPumpe3=&StatusPumpe4=&StatusPumpe5=&StatusPumpe6=&StatusPumpe7=&Taktualno=&Tfs=&Tmax=&Tmin=&Tsobna=&Tsolar=&Tspv=&Tzad_komf=&Tzad_mraz=&Tzad_red=&Tzadata=&mod_rada=&mod_rezim=&__Time=&__Date=&Jeftina_tarifa=&grejanje_off=&Alarm_tank=&Alarm_solar=&STATE_Preklopka=&SESSIONID=-1",
    "OFF": "http://192.168.1.250/isc/set_var.aspx?mod_rada=0,-1&=&SESSIONID=-1",
    "ON": "http://192.168.1.250/isc/set_var.aspx?mod_rada=1,-1&=&SESSIONID=-1",
    "GAS_OFF": "http://192.168.1.250/isc/set_var.aspx?RezimRadaPumpe4=0,-1&=&SESSIONID=-1",
    "GAS_ON": "http://192.168.1.250/isc/set_var.aspx?RezimRadaPumpe4=3,-1&=&SESSIONID=-1",
}


def prepare(url):
    global GLOBAL_UNIX_COUNTER
    GLOBAL_UNIX_COUNTER += 1
    url_with_timestamp = f"{url}&_={GLOBAL_UNIX_COUNTER}"
    request = requests.Request("GET", url_with_timestamp, headers=req_headers)
    import server

    prepared = server.main_session.prepare_request(request)
    return prepared


def prepared_url(prepared):
    return f"{prepared.method} {prepared.url}"


def send(url):
    prepared = prepare(url)
    requests_logger.write(f"{prepared_url(prepared)} --> ")

    try:
        import server

        response = server.main_session.send(prepared)
        response.raise_for_status()  # Raise an error for HTTP error codes
        requests_logger.write(f"SUCCESS\n")
    except Exception as e:
        requests_logger.write(f"FAILED ({e})\n")
        return None, e

    return response, None


def update_history(dic):

    global HISTORY_MODE
    global HISTORY_MODE_TIME_CHANGED
    global HISTORY_MODE_TIME_ON
    global HISTORY_MODE_TIME_OFF
    global HISTORY_GAS
    global HISTORY_GAS_TIME_CHANGED
    global HISTORY_GAS_TIME_ON
    global HISTORY_GAS_TIME_OFF

    global AUTO_TIMER_STARTED
    global AUTO_TIMER_STATUS
    global AUTO_GAS_STATUS

    if HISTORY_MODE is None or HISTORY_MODE != dic["mod_rada"]:
        HISTORY_MODE = dic["mod_rada"]
        HISTORY_MODE_TIME_CHANGED = time.time()

        if dic["mod_rada"]:
            HISTORY_MODE_TIME_ON = HISTORY_MODE_TIME_CHANGED
            changes_logger.write(f"mod_rada = {dic['mod_rada']}\n")
            AUTO_TIMER_STATUS = ctext_fg(COLOR_ON, f" {timestamp()} 󰐸")
        else:
            HISTORY_MODE_TIME_OFF = HISTORY_MODE_TIME_CHANGED
            e = "\n"
            p = ""
            if HISTORY_MODE_TIME_ON and HISTORY_MODE_TIME_OFF:
                elap = elapsed_str(HISTORY_MODE_TIME_OFF, HISTORY_MODE_TIME_ON)
                e = f" -- {elap}\n"
                p = f" 󱫐 {elap}\n"

            changes_logger.write(f"mod_rada = {dic['mod_rada']}{e}")
            AUTO_TIMER_STATUS = ctext_fg(COLOR_OFF, f" {timestamp()} {p}")

            if AUTO_TIMER_STARTED:
                AUTO_TIMER_STARTED = False
                AUTO_TIMER_STAUS = ctext_fg(COLOR_OFF, f"{timestamp()} 󰜺")

    if HISTORY_GAS is None or HISTORY_GAS != dic["StatusPumpe4"]:
        HISTORY_GAS = dic["StatusPumpe4"]
        HISTORY_GAS_TIME_CHANGED = time.time()

        if dic["StatusPumpe4"]:
            HISTORY_GAS_TIME_ON = HISTORY_GAS_TIME_CHANGED
            changes_logger.write(f"StatusPumpe4 = {dic['StatusPumpe4']}\n")
            AUTO_GAS_STATUS = ctext_fg(COLOR_ON, f" {timestamp()} ")
        else:
            HISTORY_GAS_TIME_OFF = HISTORY_GAS_TIME_CHANGED
            e = "\n"
            p = ""
            if HISTORY_GAS_TIME_ON and HISTORY_GAS_TIME_OFF:
                elap = elapsed_str(HISTORY_GAS_TIME_OFF, HISTORY_GAS_TIME_ON)
                e = f" -- {elap}\n"
                p = f" 󱫐 {elap}\n"

            changes_logger.write(f"StatusPumpe4 = {dic['StatusPumpe4']}{e}")
            AUTO_GAS_STATUS = ctext_fg(COLOR_OFF, f" {timestamp()} {p}")


def do_logic_timer(dic):
    global AUTO_TIMER
    global AUTO_TIMER_STARTED
    global AUTO_TIMER_SECONDS
    global AUTO_TIMER_SECONDS_ELAPSED
    global AUTO_TIMER_STATUS

    if AUTO_TIMER and int(dic["mod_rada"]):
        if AUTO_TIMER_STARTED:
            AUTO_TIMER_SECONDS_ELAPSED = time.time() - HISTORY_MODE_TIME_ON
            AUTO_TIMER_STATUS = f"{AUTO_TIMER_SECONDS_ELAPSED:.2f}/{AUTO_TIMER_SECONDS}"

            if AUTO_TIMER_SECONDS_ELAPSED >= AUTO_TIMER_SECONDS:
                AUTO_TIMER_STARTED = False
                AUTO_TIMER_STATUS = ctext_fg(COLOR_OFF, f"{timestamp()} 󱪯")
                if HISTORY_MODE_TIME_ON:
                    e = elapsed_str(time.time(), HISTORY_MODE_TIME_ON)
                    AUTO_TIMER_STATUS = ctext_fg(COLOR_OFF, f"󱫐 {e} 󱪯")
                send(URLS["OFF"])

        else:
            AUTO_TIMER_STARTED = True
            AUTO_TIMER_STATUS = ctext_fg(COLOR_ON, f"{timestamp()} 󱫌")


def do_logic_gas(dic):
    if AUTO_GAS and int(dic["StatusPumpe4"]) == 0 and dic["TminLT"]:
        AUTO_GAS_STATUS = ctext_fg(COLOR_ON, f"{timestamp()} ")
        send(URLS["GAS_ON"])

    if AUTO_GAS and int(dic["StatusPumpe4"]) == 3 and dic["TmidGE"]:
        AUTO_GAS_STATUS = ctext_fg(COLOR_OFF, f"{timestamp()} 󰙇")
        if HISTORY_GAS_TIME_ON and HISTORY_GAS_TIME_OFF:
            AUTO_GAS_STATUS = ctext_fg(COLOR_OFF, f"󱫐 {elapsed_str(time.time(), HISTORY_GAS_TIME_ON)} 󰙇")
        send(URLS["GAS_OFF"])


def remember_vars_do_action(dic):
    global AUTO_GAS
    global AUTO_GAS_STATUS

    update_history(dic)
    do_logic_timer(dic)
    do_logic_gas(dic)

    return 0


last_data = None
last_ret = False


def make_request():
    global last_data
    global last_ret

    # Send GET request
    response, err = send(URLS["VARS"])

    if response is None:
        last_ret = False
        data = last_data

        if last_data is None:
            print(f"[{timestamp()}] {err.__class__.__name__}")
            return
    else:
        data = response.json()
        last_data = data

    dic = process_data_and_draw_ui(data, last_ret, is_request=True)

    # send requests based on processed data
    remember_vars_do_action(dic)

    return dic


from server import MAIN_WORKER_DRAW_SLEEP


REQUEST_SLEEP = 3  # on third second to request
do_reqest_on_count = REQUEST_SLEEP / MAIN_WORKER_DRAW_SLEEP
request_count = 0


def do_work():

    global last_data
    global last_ret
    global request_count
    last_ret = True
    request_count += 1

    if request_count >= do_reqest_on_count or last_data is None:
        request_count = 0
        return make_request()

    dic = process_data_and_draw_ui(last_data, last_ret)
    return dic

File: src/reqworker.c
#include "reqworker.h"
#include "debug.h"
#include "drawui.h"
#include "globals.h"
#include "logger.h"
#include "mongoose.h"
#include "reqworker.h"
#include "spinners.h"
#include "utils.h"
#include <float.h>
#include <pthread.h>
#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

const char* URL_VARS = "http://192.168.1.250/isc/get_var_js.aspx?StatusPumpe3=&StatusPumpe4=&StatusPumpe5=&StatusPumpe6=&StatusPumpe7=&Taktualno=&Tfs=&Tmax=&Tmin=&Tsobna=&Tsolar=&Tspv=&Tzad_komf=&Tzad_mraz=&Tzad_red=&Tzadata=&mod_rada=&mod_rezim=&__Time=&__Date=&Jeftina_tarifa=&grejanje_off=&Alarm_tank=&Alarm_solar=&STATE_Preklopka=&SESSIONID=-1";
const char* URL_HEAT_OFF = "http://192.168.1.250/isc/set_var.aspx?mod_rada=0,-1&=&SESSIONID=-1";
const char* URL_HEAT_ON = "http://192.168.1.250/isc/set_var.aspx?mod_rada=1,-1&=&SESSIONID=-1";
const char* URL_GAS_OFF = "http://192.168.1.250/isc/set_var.aspx?RezimRadaPumpe4=0,-1&=&SESSIONID=-1";
const char* URL_GAS_ON = "http://192.168.1.250/isc/set_var.aspx?RezimRadaPumpe4=3,-1&=&SESSIONID=-1";

const char* URL_WTTRIN = "https://wttr.in/?format=4";

extern int g_running;

static char* REQUEST_FORMAT_BAS =
    "GET %s HTTP/1.0\r\n"
    "Host: %.*s\r\n"
    "Accept: application/json, text/javascript, */*; q=0.01\r\n"
    "Accept-Encoding: gzip, deflate\r\n"
    "Accept-Language: en-US,en;q=0.9\r\n"
    "Connection: keep-alive\r\n"
    "Cookie: i18next=srb\r\n"
    "Referer: http://192.168.1.250/\r\n"
    "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36\r\n"
    "X-Requested-With: XMLHttpRequest\r\n"
    "\r\n";

static char* REQUEST_FORMAT_WTTRIN =
    "GET %s HTTP/1.0\r\n"
    "Host: %.*s\r\n"
    "Content-Type: octet-stream\r\n"
    "Content-Length: 0\r\n"
    "\r\n";

static const char* s_url = NULL;
static const char* s_request_format = NULL;
#define TIMEOUT_BAS    1500
#define TIMEOUT_WTTRIN 5000
static uint64_t s_timeout_ms = TIMEOUT_BAS;

#define ERROR_NONE    0
#define ERROR_TIMEOUT 1
#define ERROR_CONN    2
static int s_errors = 0;
static int s_remember_response = 0;
static struct mg_str s_response_body = {0};

struct bas_info g_info = {0};

char g_wttrin_buffer[BIGBUFF] = {0};

long long GLOBAL_UNIX_COUNTER = 0;

atomic_int g_auto_timer;
atomic_int g_auto_gas;
atomic_int g_auto_timer_seconds;
int g_auto_timer_started = 0;
int g_auto_timer_seconds_elapsed = 0;
char g_auto_timer_status[STATUS_BUFFER_SIZE] = "...";
char g_auto_gas_status[STATUS_BUFFER_SIZE] = "...";

int g_history_mode = -1;
time_t g_history_mode_time_changed = 0;
time_t g_history_mode_time_on = 0;
time_t g_history_mode_time_off = 0;
int g_history_gas = -1;
time_t g_history_gas_time_changed = 0;
time_t g_history_gas_time_on = 0;
time_t g_history_gas_time_off = 0;

void init_reqworker()
{
    // for req
    GLOBAL_UNIX_COUNTER = timestamp();

    atomic_init(&g_auto_timer, ENABLE_AUTO_TIMER);
    atomic_init(&g_auto_gas, ENABLE_AUTO_GAS);
    atomic_init(&g_auto_timer_seconds, AUTO_TIMER_SECONDS);

    // for drawui
    init_spinners();
}

char* sendreq_error_to_str(int e)
{
    switch (e) {
        case 0:
            return "no error";
            break;
        case 1:
            return "timeout";
            break;
        case 2:
            return "connection error";
            break;
    }
    return "?";
}

// Print HTTP response and signal that we're done
static void fn(struct mg_connection* c, int ev, void* ev_data)
{
    if (ev == MG_EV_OPEN) {
        // Connection created. Store connect expiration time in c->data
        *(uint64_t*)c->data = mg_millis() + s_timeout_ms;
    }
    else if (ev == MG_EV_POLL) {
        if (mg_millis() > *(uint64_t*)c->data && (c->is_connecting || c->is_resolving)) {
            mg_error(c, "Connect timeout");
            s_errors = ERROR_TIMEOUT;
        }
    }
    else if (ev == MG_EV_CONNECT) {
        // Connected to server. Extract host name from URL
        struct mg_str host = mg_url_host(s_url);

        if (mg_url_is_ssl(s_url)) {
            struct mg_tls_opts opts = {.ca = mg_unpacked("/certs/ca.pem"), .name = mg_url_host(s_url)};
            mg_tls_init(c, &opts);
        }

        // Send request
        mg_printf(c, s_request_format, mg_url_uri(s_url), (int)host.len, host.buf);
        DPL("SENDREQ FN:");
        D(printf(s_request_format, mg_url_uri(s_url), (int)host.len, host.buf));
    }
    else if (ev == MG_EV_HTTP_MSG) {

        struct mg_http_message* hm = (struct mg_http_message*)ev_data;
        printf("%.*s", (int)hm->message.len, hm->message.buf);

        if (s_remember_response) {
            // Response received
            s_response_body = mg_strdup(hm->body);
        }

        c->is_draining = 1;        // Tell mongoose to close this connection
        *(bool*)c->fn_data = true; // Tell event loop to stop
    }
    else if (ev == MG_EV_ERROR) {
        *(bool*)c->fn_data = true; // Error, tell event loop to stop
        s_errors = ERROR_CONN;
        D(printf("CONNECTION ERROR: %d\n", s_errors));
    }
}

int sendreq(const char* url, int log, int remember_response)
{
    char request_url[REQUEST_URL_BUFFER_SIZE];
    snprintf(request_url, REQUEST_URL_BUFFER_SIZE, "%s&_=%lld", url, GLOBAL_UNIX_COUNTER);
    if (log) { logger_requests_write("%s\n", request_url); }

    s_url = url; // set url
    s_request_format = REQUEST_FORMAT_BAS;
    s_timeout_ms = TIMEOUT_BAS;
    s_remember_response = remember_response;
    s_errors = 0;                                     // RESET ERRORS
    struct mg_mgr mgr;                                // Event manager
    bool done = false;                               // Event handler flips it to true
    mg_mgr_init(&mgr);                                // Initialise event manager
    mg_http_connect(&mgr, s_url, fn, &done);          // Create client connection
    while (g_running && !done) mg_mgr_poll(&mgr, 50); // Event manager loops until 'done'
    mg_mgr_free(&mgr);                                // Free resources

    if (s_errors) {
        logger_errors_write("%s -- %s\n", request_url, sendreq_error_to_str(s_errors));
    }

    return s_errors;
}

int sendreq_wttrin(const char* url, int log, int remember_response)
{
    char request_url[REQUEST_URL_BUFFER_SIZE];
    snprintf(request_url, REQUEST_URL_BUFFER_SIZE, "%s", url);
    if (log) { logger_requests_write("%s\n", request_url); }

    s_url = url; // set url
    s_request_format = REQUEST_FORMAT_WTTRIN;
    s_timeout_ms = TIMEOUT_WTTRIN;
    s_remember_response = remember_response;
    s_errors = 0;                                     // RESET ERRORS
    struct mg_mgr mgr;                                // Event manager
    int done = 0;                                     // Event handler flips it to true
    mg_mgr_init(&mgr);                                // Initialise event manager
    mg_http_connect(&mgr, s_url, fn, &done);          // Create client connection
    while (g_running && !done) mg_mgr_poll(&mgr, 50); // Event manager loops until 'done'
    mg_mgr_free(&mgr);                                // Free resources

    if (s_errors) {
        D(printf("WTTRIN ERROR: %d\n", s_errors));
        logger_errors_write("%s -- %s\n", request_url, sendreq_error_to_str(s_errors));
    }

    return s_errors;
}

double extract(struct mg_str json_body, const char* label)
{
    D(printf("%s", json_body.buf));
    D(printf(" -- extract: %s", label));
    struct mg_str tok = mg_json_get_tok(json_body, label);

    double value = DBL_MIN;
    mg_json_get_num(tok, "$.value", &value);

    D(printf(" -> %f", value));
    return value;
}

void update_history(int mod_rada, int StatusPumpe4)
{

    DPL("UPDATE HISTORY");
    time_t current_time;
    time(&current_time);
    char* t = get_current_time();

    if (g_history_mode == -1 || g_history_mode != mod_rada) {
        g_history_mode = mod_rada;
        g_history_mode_time_changed = current_time;

        if (mod_rada) {
            g_history_mode_time_on = g_history_mode_time_changed;
            logger_changes_write("mod_rada = %d\n", mod_rada);
            snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, " %s 󰐸", t);
        }
        else {
            g_history_mode_time_off = g_history_mode_time_changed;
            char e[100] = "\n";
            char p[100] = "";
            if (g_history_mode_time_on && g_history_mode_time_off) {
                // Assuming elapsed_str() will return a string with the elapsed time
                char* elap = elapsed_str(g_history_mode_time_off, g_history_mode_time_on);
                snprintf(e, sizeof(e), " -- %s\n", elap);
                snprintf(p, sizeof(p), " 󱫐 %s", elap);
                free(elap);
            }

            logger_changes_write("mod_rada = %d%s", mod_rada, e);
            snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, " %s %s", t, p);

            if (g_auto_timer_started) {
                g_auto_timer_started = 0;
                snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, "%s 󰜺", t);
            }
        }
    }

    if (g_history_gas == -1 || g_history_gas != StatusPumpe4) {
        g_history_gas = StatusPumpe4;
        g_history_gas_time_changed = current_time;

        if (StatusPumpe4) {
            g_history_gas_time_on = g_history_gas_time_changed;
            logger_changes_write("StatusPumpe4 = %d\n", StatusPumpe4);
            snprintf(g_auto_gas_status, STATUS_BUFFER_SIZE, " %s ", t);
        }
        else {
            g_history_gas_time_off = g_history_gas_time_changed;
            char e[100] = "\n";
            char p[100] = "";
            if (g_history_gas_time_on && g_history_gas_time_off) {
                // Assuming elapsed_str() will return a string with the elapsed time
                char* elap = elapsed_str(g_history_gas_time_off, g_history_gas_time_on);
                snprintf(e, sizeof(e), " -- %s\n", elap);
                snprintf(p, sizeof(p), " 󱫐 %s", elap);
                free(elap);
            }

            logger_changes_write("StatusPumpe4 = %d%s", StatusPumpe4, e);
            sprintf(g_auto_gas_status, " %s %s", t, p);
        }
    }

    free(t);
}

void do_logic_timer(int mod_rada)
{

    time_t current_time;
    time(&current_time);

    char* t = get_current_time();

    if (g_auto_timer && mod_rada) {
        if (g_auto_timer_started) {
            g_auto_timer_seconds_elapsed = difftime(current_time, g_history_mode_time_on);
            snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, "%d/%d", g_auto_timer_seconds_elapsed, atomic_load(&g_auto_timer_seconds));

            if (g_auto_timer_seconds_elapsed >= g_auto_timer_seconds) {
                g_auto_timer_started = 0;
                snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, "%s 󱪯", t);
                if (g_history_mode_time_on) {
                    char* elap = elapsed_str(time(NULL), g_history_mode_time_on);
                    snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, "󱫐 %s 󱪯", elap);
                    free(elap);
                }
                sendreq(URL_HEAT_OFF, 1, 0);
            }
        }
        else {
            g_auto_timer_started = 1;
            snprintf(g_auto_timer_status, STATUS_BUFFER_SIZE, "%s 󱫌", t);
        }
    }

    free(t);
}

void do_logic_gas(int StatusPumpe4, int TminLT, int TmidGE)
{

    char* t = get_current_time();

    if (g_auto_gas && StatusPumpe4 == 0 && TminLT) {
        sprintf(g_auto_gas_status, "%s ", t);
        sendreq(URL_GAS_ON, 1, 0);
    }

    if (g_auto_gas && StatusPumpe4 == 3 && TmidGE) {
        sprintf(g_auto_gas_status, "%s 󰙇", t);
        if (g_history_gas_time_on && g_history_gas_time_off) {
            char* elap = elapsed_str(time(NULL), g_history_gas_time_on);
            sprintf(g_auto_gas_status, "󱫐 %s 󰙇", elap);
            free(elap);
        }
        sendreq(URL_GAS_OFF, 1, 0);
    }

    free(t);
}

void remember_vars_do_action(int mod_rada, int StatusPumpe4, int TminLT, int TmidGE)
{
    update_history(mod_rada, StatusPumpe4);
    do_logic_timer(mod_rada);
    do_logic_gas(StatusPumpe4, TminLT, TmidGE);
}

extern double g_temp_buf_max;
extern double g_temp_buf_min;

void update_info()
{
    GLOBAL_UNIX_COUNTER++;

    // get request, parse response
    sendreq(URL_VARS, 0, 1);
    if (s_response_body.buf) {
        D(printf("RESPONSE BODY BUF LEN: %lu\n", strlen(s_response_body.buf)));
        g_info.hasValues = 1;
        g_info.mod_rada = extract(s_response_body, "$.mod_rada");
        g_info.mod_rezim = extract(s_response_body, "$.mod_rezim");
        g_info.StatusPumpe3 = extract(s_response_body, "$.StatusPumpe3");
        g_info.StatusPumpe4 = extract(s_response_body, "$.StatusPumpe4");
        g_info.StatusPumpe5 = extract(s_response_body, "$.StatusPumpe5");
        g_info.StatusPumpe6 = extract(s_response_body, "$.StatusPumpe6");
        g_info.StatusPumpe7 = extract(s_response_body, "$.StatusPumpe7");
        g_info.Tspv = extract(s_response_body, "$.Tspv");
        g_info.Tsolar = extract(s_response_body, "$.Tsolar");
        g_info.Tzadata = extract(s_response_body, "$.Tzadata");
        g_info.Tfs = extract(s_response_body, "$.Tfs");
        g_info.Tmax = extract(s_response_body, "$.Tmax");
        g_info.Tmin = extract(s_response_body, "$.Tmin");
        g_info.Tsobna = extract(s_response_body, "$.Tsobna");

        // calc other values
        g_info.Tmid = (g_info.Tmax + g_info.Tmin) / 2;
        g_info.Thottest = g_temp_buf_max;
        g_info.Tcoldest = g_temp_buf_min;
        g_info.TminLT = g_info.Tmin < 45;
        g_info.TmidGE = g_info.Tmid >= 60;

        // free buffer
        free((void*)s_response_body.buf);
        s_response_body.buf = NULL;
    }

    draw_ui(g_info, 1, s_errors);

    remember_vars_do_action(g_info.mod_rada, g_info.StatusPumpe4, g_info.TminLT, g_info.TmidGE);
}

void wttrin_get_weather()
{
    DPL("WTTRIN SENDREQ");
    sendreq_wttrin(URL_WTTRIN, 0, 1);

    if (s_response_body.buf) {
        D(printf("WTTRIN RESPONSE BODY BUF LEN: %lu\n", strlen(s_response_body.buf)));

        // write response to buffer
        snprintf(g_wttrin_buffer, BIGBUFF, "%s", s_response_body.buf);

        // free buffer
        free((void*)s_response_body.buf);
        s_response_body.buf = NULL;
    }

    exit(1);
}

static int request_count = 0;
static int wttrin_request_count = 0;

void reqworker_do_work()
{
    request_count++;
    wttrin_request_count++;

    if (!g_info.hasValues || request_count >= DO_REQUEST_COUNT) {
        request_count = 0;
        update_info();
    }

    if (g_wttrin_buffer[0] == 0 || wttrin_request_count >= WTTRIN_DO_REQUEST_COUNT) {
        wttrin_request_count = 0;
        wttrin_get_weather();
    }

    draw_ui(g_info, 0, s_errors);
}
